src/luarocks/fetch/git.lua

```diff
@@ -0,0 +1,142 @@
+local function git_can_clone_by_tag(git_cmd)
+  local version_string = io.popen(fs.Q(git_cmd)..' --version'):read()
+  local major, minor, tiny = version_string:match('(%d-)%.(%d+)%.?(%d*)')
+  major, minor, tiny = tonumber(major), tonumber(minor), tonumber(tiny) or 0
+  local value = major > 1 or (major == 1 and (minor > 7 or (minor == 7 and tiny >= 10)))
+  git_can_clone_by_tag = function() return value end
+  return value
+end
+
+
+local function git_commit(git_cmd, path)
+  local command = {
+    fs.Q(git_cmd),
+    "--git-dir", fs.Q(dir.path(path, '.git')),
+    '--work-tree', fs.Q(path),
+    'rev-parse', '--verify', 'HEAD'
+  }
+
+  local fd = io.popen(table.concat(command, ' '))
+  local out = fd:read("*l")
+  local ok = fd:close()
+
+  if ok then
+    return out
+  else
+    return nil, 'could not get git commit hash'
+  end
+end
+
+local function update_rockspec(rockspec, changes)
+  local file = rockspec.local_filename
+  local cached = persist.load_into_table(file)
+
+  util.deep_merge(cached, changes)
+  util.deep_merge(rockspec, changes)
+
+  return persist.save_from_table(file, cached, type_check.rockspec_order)
+end
+--- Download sources for building a rock, using git.
+-- @param rockspec table: The rockspec table
+-- @param extract boolean: Unused in this module (required for API purposes.)
+-- @param dest_dir string or nil: If set, will extract to the given directory.
+-- @return (string, string) or (nil, string): The absolute pathname of
+-- the fetched source tarball and the temporary directory created to
+-- store it; or nil and an error message.
+function git.get_sources(rockspec, extract, dest_dir, depth)
+  assert(type(rockspec) == "table")
+  assert(type(dest_dir) == "string" or not dest_dir)
+
+  local git_cmd = rockspec.variables.GIT
+  local name_version = rockspec.name .. "-" .. rockspec.version
+  local module = dir.base_name(rockspec.source.url)
+  -- Strip off .git from base name if present
+  module = module:gsub("%.git$", "")
+
+  local ok, err_msg = fs.is_tool_available(git_cmd, "Git")
+  if not ok then
+    return nil, err_msg
+  end
+
+  local store_dir
+  if not dest_dir then
+    store_dir = fs.make_temp_dir(name_version)
+    if not store_dir then
+      return nil, "Failed creating temporary directory."
+    end
+    util.schedule_function(fs.delete, store_dir)
+  else
+    store_dir = dest_dir
+  end
+  store_dir = fs.absolute_name(store_dir)
+  local ok, err = fs.change_dir(store_dir)
+  if not ok then return nil, err end
+
+  local commit = string.match(rockspec.source.hash or '', '%w+')
+  if not depth and commit then depth = '--shallow-submodules' end -- setting it to empty string fails
+
+  local command = {fs.Q(git_cmd), "clone", depth or "--depth=1", rockspec.source.url, module}
+  local tag_or_branch = rockspec.source.tag or rockspec.source.branch
+  -- If the tag or branch is explicitly set to "master" in the rockspec, then
+  -- we can avoid passing it to Git since it's the default.
```

Comment:
I am not really sure of that
Can’t we choose a default branch in GH and when we clone it will be this branch that will be default?

https://stackoverflow.com/questions/18726037/what-determines-default-branch-after-git-clone

Still depends on the version of git >1.8.5
Maybe safer/simpler to add the branch?
Also if you know what you are cloning you can use `—depth=1` So you are not pulling the whole history of the branch

Or maybe using this https://stackoverflow.com/a/3489576

Reply:
This code is copied from luarocks itself with no changes. Honestly I don't really care much what it does because rover in later commits overrides it to just clone whole repo and do a reset.